File: Typescript/twoSum.ts
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = (nums: number[], target: number): number[] => {
    const numsHash: { [key: number]: number } = {};
    for (let i = 0; i < nums.length; ++i) {
        const complement = target - nums[i];
        if (numsHash[complement] !== undefined) {
            return [numsHash[complement], i];
        }
        numsHash[nums[i]] = i;
    }
    return [-1, -1];
};

File: C++/twoSum.cpp
class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> numsHashTable;

        for (int i = 0; i < nums.size(); ++i) {
            const int val = target - nums[i];
            const auto itemIt = numsHashTable.find(val);

            if (itemIt != numsHashTable.end()) {
                return {itemIt->second, i};
            }

            numsHashTable[nums[i]] = i;
        }

        return {-1, -1};
    }
};
